genie-core/src/main/java/com/netflix/genie/core/services/impl/JobCoordinatorServiceImpl.java

```diff
@@ -239,60 +211,43 @@ public String coordinateJob(
-
-                            // Tell the system a new job has been scheduled so any actions can be taken
-                            log.info("Publishing job scheduled event for job {}", jobId);
-                            this.eventPublisher.publishEvent(new JobScheduledEvent(jobId, task, memory, this));
-                            return jobId;
-                        } catch (final TaskRejectedException e) {
-                            throw new GenieServerException(
-                                "Unable to launch job due to exception: " + e.getMessage(),
-                                e
-                            );
-                        }
-                    } else {
-                        throw new GenieServerUnavailableException(
-                            "Job "
-                                + jobId
-                                + " can't run on this node "
-                                + usedMemory
-                                + "/"
-                                + maxSystemMemory
-                                + " MB are used and requested "
-                                + memory
-                                + " MB"
-                        );
-                    }
-                } catch (final Exception e) {
-                    this.jobPersistenceService.updateJobStatus(jobId, JobStatus.FAILED, e.getMessage());
-                    throw e;
+                log.info("Checking if can run job {} on this node", jobRequest.getId());
+                final int maxSystemMemory = this.jobsProperties.getMemory().getMaxSystemMemory();
+                final int usedMemory = this.jobStateService.getUsedMemory();
+                if (usedMemory + memory <= maxSystemMemory) {
+                    log.info(
+                        "Job {} can run on this node as only {}/{} MB are used and requested {} MB",
+                        jobId,
+                        usedMemory,
+                        maxSystemMemory,
+                        memory
+                    );
+                    // Tell the system a new job has been scheduled so any actions can be taken
+                    log.info("Publishing job scheduled event for job {}", jobId);
+                    jobStateService.schedule(jobId, jobRequest, cluster, command, applications, memory);
+                    return jobId;
+                } else {
+                    throw new GenieServerUnavailableException(
+                        "Job "
+                            + jobId
+                            + " can't run on this node "
+                            + usedMemory
+                            + "/"
+                            + maxSystemMemory
+                            + " MB are used and requested "
+                            + memory
+                            + " MB"
+                    );
                 }
             }
+        } catch (GenieException e) {
+            jobStateService.done(jobId);
```

Comment:
It does makes sense but wanted to keep it separate to being with. First draft to be in-memory. This should actually have all the job related persistence steps done from within here.